flask/testing.py

```diff
@@ -45,41 +44,72 @@ def make_test_environ_builder(
         :class:`~werkzeug.test.EnvironBuilder`.
     """
 
-    assert not (base_url or subdomain or url_scheme) or (base_url is not None) != bool(
-        subdomain or url_scheme
-    ), 'Cannot pass "subdomain" or "url_scheme" with "base_url".'
+    def __init__(
+        self,
+        app,
+        path="/",
+        base_url=None,
+        subdomain=None,
+        url_scheme=None,
+        *args,
+        **kwargs
+    ):
+        assert not (base_url or subdomain or url_scheme) or (
+            base_url is not None
+        ) != bool(
+            subdomain or url_scheme
+        ), 'Cannot pass "subdomain" or "url_scheme" with "base_url".'
+
+        if base_url is None:
+            http_host = app.config.get("SERVER_NAME") or "localhost"
+            app_root = app.config["APPLICATION_ROOT"]
+
+            if subdomain:
+                http_host = "{0}.{1}".format(subdomain, http_host)
+
+            if url_scheme is None:
+                url_scheme = app.config["PREFERRED_URL_SCHEME"]
+
+            url = url_parse(path)
+            base_url = "{scheme}://{netloc}/{path}".format(
+                scheme=url.scheme or url_scheme,
+                netloc=url.netloc or http_host,
+                path=app_root.lstrip("/"),
+            )
+            path = url.path
 
-    if base_url is None:
-        http_host = app.config.get("SERVER_NAME") or "localhost"
-        app_root = app.config["APPLICATION_ROOT"]
+            if url.query:
+                sep = b"?" if isinstance(url.query, bytes) else "?"
+                path += sep + url.query
 
-        if subdomain:
-            http_host = "{0}.{1}".format(subdomain, http_host)
+        if "json" in kwargs:
```

Comment:
Should be able to remove this part, since the base `EnvironBuilder` handles JSON.

Reply:
No, because the base class doesn't know about `app` and therefore does not take JSON serialisation settings from it.

This could be removed if `json_dumps()` was added as a method that respects `self.app`, but this is hard if json_dumps() needs to be callable as a static method. If that's not needed, then yes, I can remove this.

Reply:
In fact, I'm pretty sure it is a safe assumption that it's not needed. Let me go ahead and do this.

Reply:
Yeah, that should work.